src/Razor/Microsoft.AspNetCore.Mvc.Razor.Extensions/src/ViewComponentTagHelperDescriptorProvider.cs

```diff
@@ -39,6 +39,18 @@ public void Execute(TagHelperDescriptorProviderContext context)
             var visitor = new ViewComponentTypeVisitor(vcAttribute, nonVCAttribute, types);
             var discoveryMode = context.Items.GetTagHelperDiscoveryFilter();
 
+            if ((discoveryMode & TagHelperDiscoveryFilter.TargetAssembly) == TagHelperDiscoveryFilter.TargetAssembly)
+            {
+                var targetReference = context.Items.GetTargetMetadataReference();
+                if (targetReference is IAssemblySymbol assembly)
+                {
+                    if (IsTagHelperAssembly(assembly))
+                    {
+                        visitor.Visit(assembly.GlobalNamespace);
+                    }
+                }
+            }
+
             if ((discoveryMode & TagHelperDiscoveryFilter.CurrentCompilation) == TagHelperDiscoveryFilter.CurrentCompilation)
```

Comment:
Would we still want to keep the notion of filtering? We could use the presence of `GetTargetMetadataReference` to determine if it walks everything vs one assembly and returns, no?

Reply:
Yep -- that's actually where I started off with this PR but i felt that leveraging the notion of filtering made the implementation a lot more explicit and easier to read.

Reply:
The flag has to be mutually exclusive with the current filtering option otherwise you end up with duplicate descriptors. More importantly, nothing would use the other two options which seems like a good reason to not have it any longer.

Reply:
Updated this based on our Teams convo.